### ISSUE TITLE
Update HTTP signatures to support IETF draft from October 2021

### ISSUE THREAD

**Issue by ruhrscholz** [mastodon/mastodon#21429]
### Pitch

The current Mastodon implementation supports rsa-sha256 and hs2019 as valid HTTP signature algorithms:

https://github.com/mastodon/mastodon/blob/8cf7006d4efbcfdd4a4ab688db1bcc73a2915a47/app/controllers/concerns/signature_verification.rb#L83

At the time of implementation those were the valid algorithms in the IETF draft. However, since then, in October 2021 (revision 03), a list of new algorithms was added: https://www.ietf.org/rfcdiff?difftype=--hwdiff&url2=draft-ietf-httpbis-message-signatures-03.txt
Furthermore, in revision 08, EdDSA (ed25519) was added: https://www.ietf.org/rfcdiff?difftype=--hwdiff&url2=draft-ietf-httpbis-message-signatures-08.txt
For reference, the current (at the time of writing) IETF draft is revision 15: https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-message-signatures-15

### Motivation

Even though the current implementation is technically valid as per Chapter 1.4 ("For example, a server expecting only ECDSA signatures should know to reject any RSA signatures"), support for more signature algorithms could be added to improve intercompatibility and to better adhere to the current specs.

**Comment by ruhrscholz** [comment]
The HTTP Message Signatures are the official updated replacement for Mastodon's HTTP Signatures. See https://datatracker.ietf.org/doc/draft-cavage-http-signatures/12/, field "Replaced By"

**Comment by nightpool** [comment]
Thanks for filing this ticket. Can you shed some light into motivation for this change besides just supporting more formats for the sake of supporting more formats? Is there a particularly use-case that supporting ed25519 would unlock for you? As you can imagine, there's been a lot of churn in the HTTP Signatures spec lately, and it would be helpful to have any info to help us prioritize these changes. Additionally, every new signature scheme supported opens up new complexity in terms of key parsing and possible key confusion attacks, so unless there's a compelling benefit I'm not sure that "support every signature scheme in the spec" is necessarily a good goal

**Comment by ruhrscholz** [comment]
My main motivation stems from the fact that in a lot of different use cases, most modern solutions use Ed25519 (or other elliptic curves) over RSA. The main example I get to use on an every-day basis are SSH keys. Even [GitHub recommends the use of Ed25519](https://docs.github.com/en/authentication/connecting-to-github-with-ssh/generating-a-new-ssh-key-and-adding-it-to-the-ssh-agent#generating-a-new-ssh-key) for creating new keys. It also is in use in the [US government](https://web.archive.org/web/20090207005135/http://www.nsa.gov/ia/programs/suiteb_cryptography/index.shtml), the [Tor protocol](https://gitweb.torproject.org/tor.git/tree/ReleaseNotes?h=release-0.2.4), [BitCoin](https://en.bitcoin.it/wiki/Elliptic_Curve_Digital_Signature_Algorithm), [iMessage](https://blog.quarkslab.com/imessage-privacy.html), and a probably lot more. This means there is wide support in every semi-modern environment I could think of. Mastodon already uses OpenSSL to verify the RSA signatures - OpenSSL also has battle-proven ECC implementations. Since in a while ago the `algorithm` field has been dropped from HTTP signatures (see #7520), the implementation should be even less work and I believe a part of it could be offloaded to the OpenSSL library itself, but I'm not fluent enough in Ruby to figure that out.

The advantages in using ECC do not consist of new features or use cases, but are more about efficiency, space requirements and thus handling more users per hardware unit:

Space requirements: Elliptic curve cryptography delivers the same level of resistance as RSA with lower key length, which when caching many users keys pays off from a database perspective. Per NIST and some independent studies, a 2048bit RSA key is roughly equivalent to an 224bit ECC key (see https://www.globalsign.com/en/blog/elliptic-curve-cryptography, and https://eprint.iacr.org/2013/635.pdf)

Efficiency: Elliptic curve cryptography is easier to scale up, since it [requires less computing power](https://ww1.microchip.com/downloads/en/DeviceDoc/00003442A.pdf) to verify a given signature (this stems from the smaller key sizes). Less CPU consumption (and thus power consumption) on big instances.

Implementation: I'm definitely not an expert on the math behind those two, but from what I could gather, ECC is simpler and has less pitfalls when implementing, but when relying on a battle-proven library, this should not matter that much.

Cryptography: There are some purely-mathematic advantages in using ECC, but I have to admit I'm not smart enough to understand them. The [Wikipedia article](https://en.wikipedia.org/wiki/Elliptic-curve_cryptography) is a nice starting point.

There are other advantages in other use cases, like [forward secrecy](https://blog.cloudflare.com/staying-on-top-of-tls-attacks/) when used in Diffie-Hellman Key Exchanges, which lead to a greater adoption overall but do not matter as much when signing messages like in our use-case.

Regarding our domain-specific (ActivityPub) use case: The HTTP Signature spec seems to be close to being finished, and from my perspective the missing signature algorithms could become a major source of incompatibility. Even if Mastodon does not use ECC to sign its own requests, a lot of other implementation may do and probably will do for the reasons mentioned above - and since it is an official spec. So their either will be a plain incompatibility or fallback attempts: Try ECC, get 401, try RSA, effectively doubling the number of requests against a give Mastodon instance. Apart from that, a number of issues regarding outdated spec-compliance have already been fixed: #17004 #12914 #7520

**Comment by janbruns** [comment]
To me, it doesn't look like a simple "update". It not a specification revision, but an entirely different and incompatible method.

Using algo "hs2019", it might also be possible to use ECC within the older cavage-http-signatures specification and without any code change: "hs2019" delegates definition of the asym signature algo part to the keyId meta data, and given mastodon makes use of PEM data, it already has some "type of key"=>"signing algo" information, and assuming it makes use of OpenSSL's higher level functions, everything might already get autodetected.

**Comment by ruhrscholz** [comment]
Thank you for the update, while reading the drafts I did not realize that HS2019 is a container which can include different signature algorithms. Upon re-reading them, it becomes more clear. I initially thought it was just some alternative to RSA/ECC.

Apart from that, what is the general stance on HTTP Signatures? Will the Mastodon project stick to the newest "cavage" draft or will there still be updates in the future?

**Comment by nightpool** [comment]
I'm not sure, we'll have to review the changes and see how/if they can be made backwards compatibly

**Comment by mariusor** [comment]
@nightpool, regarding specific use cases for ed25519, I am developing an [ActivityPub server](https://github.com/go-ap/fedbox) that uses them for its actors and currently it can't interact with Mastodon because of that.

I would like to add hs2019 support for encoding the keys, but currently there are no libraries that I could find that do this work in my language of choice (Go), so I would have to write my own.

**Comment by janbruns** [comment]
Maybe probs like this?
https://github.com/openssl/openssl/issues/11571

**Comment by mariusor** [comment]
@ruhrscholz @janbruns I just realized that using HS2019 is not going to work for signatures generated based on keys that are not RSA256. In this [message](https://github.com/mastodon/mastodon/pull/14556#issue-676210587) it's explicitly said (emphasis mine) that:

> * add support for the hs2019 signature algorithm (**assumed to be equivalent to RSA-SHA256**, since we do not have a mechanism to specify the algorithm within the key metadata yet)

So we're back to square one.

**Comment by nightpool** [comment]
yes, that makes sense. It sounds like ed25519 support would be good to add, however, I'm not super enthused by the number and diversity of protocols in the recent spec draft and I don't think we should aim to support all of them. We'll also need to decide on a way of communicating the key metadata out of band, possibly by adding a new property, or possibly by moving to JWKs to simplify the encoding process.

Do you have thoughts here? My sense is that there's probably some existing vocabulary items in the security spec that we could use here, which probably makes more sense since it would keep us consistent, but I know there are reasons to distrust the whole existing PEM encoding mess

**Comment by mariusor** [comment]
> Do you have thoughts here?

The way I implemented public key loading and decoding in my application was by iterating through a list of algorithms (which currently holds ed2559, rsa512, rsa256) and trying to decode with each of them in turn. I am not aware of better solutions than that, sadly.

**Comment by janbruns** [comment]
Just run a cmd like
`openssl asn1parse -in public.pem`
to see that the public key pem file already includes almost all information to define signature operation, just as defined in 
https://www.rfc-editor.org/rfc/rfc7468#section-13

So commands like
`openssl dgst -sha256 -sign private.pem -out signature.bin msg.txt
openssl dgst -sha256 -verify public.pem -signature signature.bin msg.txt
`
can work fully specified without specifying such details again. For example, the public pem can contain a "rsassaPss" key instead of the older standard pkcs_v1_5 "rsaEncryption" key.

The openssl version installed on my computer however doesn't support ED25519 keys to be used that way, probably because the algo isn't yet defined within RFC7468=>RFC5280=>[RFC3279,RFC4055,RFC4491].

**Comment by janbruns** [comment]
Reinvent the wheel, duplicate all the well-established industry standard crypto stuff, just because Annabelle Backman @amazon had the idea of creating work by making things incompatible?

And all this, after mastodon already came up with it's very own ideas about object expansion (deliver almost fully expanded content data), so that it "needed" one more, again very similar, but totally different activity-Signature crypto implementation (that won't work, if  a targeted server refuses to let GIGs of LD-processing algos fetch context data from the web)?

**Comment by mariusor** [comment]
Thank you for the constructive feedback you provided so far but you are very rambly in this message, and I can't understand what you're up against. Plus, I can't understand why you feel like it's OK to single out and shame people that you disagree with.

I don't see anything in the httpsig spec that makes things incompatible with "standard crypto stuff", but I am not interested in debating with you so I'll remove myself from the conversation while mastodon devs decide if it's worth their time pursuing this issue or not. Cheers.

**Comment by janbruns** [comment]
The point is, not any specification can be expected to become a standard, but this was your claim:

"The "decision", if it can be called that, has already been made with the update to the HTTP Signatures specification."

**Comment by mariusor** [comment]
Getting back to the point, I wonder if it wouldn't be enough to change this:

https://github.com/mastodon/mastodon/blob/8cf7006d4efbcfdd4a4ab688db1bcc73a2915a47/spec/models/account_spec.rb#L233

to 

```ruby
      expect(account.keypair).to be_instance_of OpenSSL::PKey::PKey
```

The docs [seem to indicate](https://ruby-doc.org/3.1.3/exts/openssl/OpenSSL/PKey/PKey.html) that it's a wrapper for the discrete implementations of DSA, RSA and EC keys.

I couldn't wrap my head around how the remote actors have their `keypair` initialized though, to double check if it explicitly tries to read a `OpenSSL::PKey::RSA` or `OpenSSL::PKey::PKey`.

**Comment by mariusor** [comment]
Like Jenbruns showed higher in the thread, there's no need to store the type of the key, the openssl libray is capable of using the function `PKey.read` to read a public key:

```ruby
require 'openssl'

RSA = '-----BEGIN PUBLIC KEY-----
MFwwDQYJKoZIhvcNAQEBBQADSwAwSAJBAN/ovAmhoNOnE2VoJeUunpEB+CwTyPbKI8VnQ+rLa75riBtmfYrGdUD8Cv/p+ioL86xwSvX+HYXnkPffEXc7IlUCAwEAAQ==
-----END PUBLIC KEY-----
'
ED25519 = '-----BEGIN PUBLIC KEY-----
MCowBQYDK2VwAyEANYl9n9aoamu+aix442RKARnR/UWwSZ1FFf24TPe43zo=
-----END PUBLIC KEY-----
'

key = OpenSSL::PKey.read(RSA)
printf("#{key.class}\n")
# OpenSSL::PKey::RSA

key = OpenSSL::PKey.read(ED25519)
printf("#{key.class}\n")
# OpenSSL::PKey::PKey

```

**Comment by trwnh** [comment]
so something like this then?

OpenSSL::PKey.read(actor.public_key).verify(...)

**Comment by mariusor** [comment]
Yes, exactly like that :). 

The only caveat that would require differentiating between key types, taken from the man page of EVP_DigestVerify(3):

```
       Ed25519 and Ed448
           Support no digests (the digest type must be NULL)
```

This information can be obtained by checking the `key.oid` that can be `rsaEncryption` or `ED25519` for our cases.

**Comment by janbruns** [comment]
Given the relaxed security constraints for fed servers (where many aspects  can either be a community-decision or key-owner/server instance admin choice; we intentionally leave things open), maybe even messing with oids can be avoided, like:
https://jan.abnuto.de/code/httpsig_with_ed25519.txt

**Comment by aumetra** [comment]
Are there plans to eventually support schemes such as Ed25519 or Ed448 in the future?  
Both signature schemes require smaller keys, and are (especially with parallel formulas) much faster than RSA (Ed25519 would also open up the possibility of batch verification but that seems like a more niche use-case)

Kitsune [already implements support for Ed25519 signatures](https://github.com/kitsune-soc/kitsune/blob/40ee7d0c4f919827abc55abb180fc43277fd8e2d/kitsune/src/http/extractor/signed_activity.rs#L100-L112). The general format of the keys stays the same (an SPKI structure) and the algorithm is determined via the OID field in the public key in our implementation (we still need to add tests).

**Comment by silverpill** [comment]
With FEP-521a an upgrade path for Fediverse software would look like this:
1. Support FEP-521a
2. Generate Ed25519 keys for actors and attach them alongside with RSA keys.
3. Support verifying EdDSA signatures
4. Start using EdDSA for signing outgoing requests.

Compatibility with legacy software only breaks at (4)

**Comment by renchap** [comment]
The Social WG published a document on this topic, with some strategies about how to upgrade without breaking compatibility: https://swicg.github.io/activitypub-http-signature/

This is on our roadmap, but no idea when we will have some bandwidth for this.

Internal reference: MAS-215

**Comment by renchap** [comment]
As implied in the message above, we do not know when we will be able to have a go at this, neither how we will proceed.

**Comment by ThisIsMissEm** [comment]
@silverpill for "Start using EdDSA for signing outgoing requests." is there a mechanism through which an ActivityPub Server could discover if a target server/actor supports EdDSA instead of RSA? e.g., when preparing to send, you know where you're sending to, I guess, so I'm wondering if it'd be possible to discover whether it supports EdDSA or just RSA?

**Comment by silverpill** [comment]
Such mechanism has not been proposed yet. If a target actor document contains an Ed25519 public key, that can be interpreted as a signal of EdDSA support. However, it is not reliable, because server may support EdDSA in integrity proofs, but not in HTTP signatures.
I would use an additional attribute to signal support for EdDSA + HTTP signatures (this attribute can be added to an instance actor too).

**Comment by ThisIsMissEm** [comment]
@silverpill having something that solves that sounds like it'd be a good idea, as it'd prevent leaving people behind whilst everyone upgrades

**Comment by soatok** [comment]
Hi everybody. I work in cryptography. I wanted to chime into this thread with some constructive feedback.

@janbruns ([comment](#issuecomment-1328980503)):

> Using algo "hs2019", it might also be possible to use ECC within the older cavage-http-signatures specification and without any code change: "hs2019" delegates definition of the asym signature algo part to the keyId meta data,

This sound _suspiciously_ like "hs2019 is trivially vulnerable to Algorithm Confusion if an attacker tampers with the metadata".

If you aren't familiar with this class of issues, consider how [RSA/HMAC confusion breaks JWT](https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/#:~:text=Let%27s%20keep%20digging.-,RSA%20or%20HMAC%3F,-The%20JWT%20spec). The failure mode of "RSA/HMAC" confusion is clearly worse than "RSA/ECDSA" confusion, but it's still alarming for any system to just blithely accept such a confusion as part of its standard operation.

As SwiftOnSecurity puts it: [Cryptography is nightmare magic math that cares what color pen you use](https://soatok.blog/wp-content/uploads/2024/07/SwiftOnSecurity-Nightmare-Magic-Math-Cryptography.pdf).

@aumeta ([comment](#issuecomment-1549405785)):

> Kitsune [already implements support for Ed25519 signatures](https://github.com/kitsune-soc/kitsune/blob/40ee7d0c4f919827abc55abb180fc43277fd8e2d/kitsune/src/http/extractor/signed_activity.rs#L100-L112) (and would like to fully migrate to Ed25519 in the future). The general format of the keys stays the same (an SPKI structure) and the algorithm is determined via the OID field in the public key in our implementation (we still need to add tests).

Good on Kitsune for sensible defaults.

----

Although these deficits have been public for years and it would be silly to treat HS2019 as a possible security issue, fixing this should be prioritized.

**Comment by soatok** [comment]
(BTW, feel free to assign this to me. I'm going to cut a PR for this once I'm not preoccupied with $employer.)

**Comment by oneiros** [comment]
Thank you very much for your offer to help out, we really appreciate it. Sadly, I am not sure what exactly you are proposing.

The discussion in this issue seems to mix two related but ultimately seperate things: Supporting newer versions of HTTP signatures and changing the types of public keys we use in Mastodon. The former is something that is on our immediate roadmap and that we would like to track in #29905. There is still some uncertainty around the latter which means I cannot promise that we can accept a PR for this.

**Comment by silverpill** [comment]
Mastodon now implements RFC-9421, but are there any plans to support Ed25519 keys?

FEP-521a has been finalized recently: https://codeberg.org/fediverse/fep/src/branch/main/fep/521a/fep-521a.md. Related W3C specifications reached the status of Recommendation.

@ThisIsMissEm regarding your question:

> is there a mechanism through which an ActivityPub Server could discover if a target server/actor supports EdDSA instead of RSA

I published a new FEP about capability discovery: https://codeberg.org/fediverse/fep/src/branch/main/fep/844e/fep-844e.md. Capabilities are identified by URIs there, and the proposed identifier for the combination of RFC-9421 and Ed25519 is `https://datatracker.ietf.org/doc/html/rfc9421#name-eddsa-using-curve-edwards25`.

**Comment by oneiros** [comment]
> Mastodon now implements RFC-9421, but are there any plans to support Ed25519 keys?

To be precise, we merged experimental support to accept RFC9421 signatures, but the feature is still behind a flag.

The plan is roughly to
1. remove the feature flag so that we accept the new signatures and once this has been somewhat widely deployed
2. start signing outgoing requests with RFC9421 signatures
3. Add support for additional key types (and start migrating Mastodon away from the old RSA ones).

So yes, there are plans, but some work needs to be finished first.

**Comment by soatok** [comment]
Now that 4.5 is in its first Release Candidate, has 4.4 been rolled out widely enough to move forward on RFC 9421?

How can I help make this happen?

**Comment by oneiros** [comment]
To be clear: We only removed the feature flag mentioned above a few days ago. So 4.5 will be the first version to accept RFC 9421 signatures without admins having to enable this explicitly.

**Comment by soatok** [comment]
@oneiros Wonderful. I noticed the actor has a single `publicKey` object attached to it. Is there a plan to make this an array of objects in the future, or is the "multiple keys" support defined in another issue that GitHub's search isn't easily locating?

**Comment by oneiros** [comment]
As noted above, this is step three. Once we can ingest HTTP Message Signatures (done) *and* produce those signatures (not done), we can start thinking about how to migrate to using more modern key types.

We have not thought this through yet, but I suspect this will require multiple keys per actor. And currently the best way to achieve this, as far as I am aware, is [FEP-521a](https://codeberg.org/fediverse/fep/src/branch/main/fep/521a/fep-521a.md).

I cannot promise 100% that this is what we will end up with, but it seems like the most likely outcome from what I know today.

**Comment by soatok** [comment]
> Once we can ingest HTTP Message Signatures (done) _and_ produce those signatures (not done), we can start thinking about how to migrate to using more modern key types.

Great. Do you want help getting this implemented?

I understand you cannot move fast on this, but I want to make sure this feature isn't blocked by someone needing to roll up their sleeves and write the code. :3

**Comment by mradcliffe** [comment]
> I understand you cannot move fast on this, but I want to make sure this feature isn't blocked by someone needing to roll up their sleeves and write the code. :3

I think the next step would be to implement [Linzer](https://github.com/nomadium/linzer/) signatures using a double-knocking pattern and handle errors from implementations that do not support RFC 9421 yet (or ever). I would guess there would be some architectural/performance-related decisions about how best to do that in Mastodon.

From my own notes, these are the results when software has not implemented yet:

- Lemmy responds with 400 Bad Request when RFC 9421 fails.
- GoToSocial responds with 401 Unauthorized (`{"error":"Unauthorized: http request wasn't signed or http signature was invalid: (verifier)"}` when RFC 9421 fails.
- Friendica and ktistec seem to bypass signature verification completely and accept the post (false positive success).
- Mastodon responds with 401 (`{"error": "Error parsing signature parameters"}`) when the instance does not support (or has not enabled) RFC 9421.
- I forgot to jot down which software responds with (`{"error": "sig missing keyId"}`) when it doesn't support RFC 9421.

**Comment by ThisIsMissEm** [comment]
@mradcliffe that sounds like something the [ActivityPub HTTP Signatures](https://swicg.github.io/activitypub-http-signature/) document should actually specify, and implementations should align with that..

**Comment by oneiros** [comment]
> Do you want help getting this implemented?

Thanks for offering! I would certainly like your input / review when we get to do this.

**Comment by soatok** [comment]
@oneiros Certainly. Happy to help.

**Comment by no-usernames-left** [comment]
> I would certainly like your input / review when we get to do this.

Are... are you seriously declining an offer from _someone like @soatok_ to implement this for you?!

What a time to be alive.

**Comment by robey** [comment]
You may want to read/participate in the activitypub discussion on RFC 9421 migration too: https://socialhub.activitypub.rocks/t/rfc-9421-http-signatures-in-2026/8427

In particular, it looks like servers can advertise support with an HTTP header ("Accept-signature") which may solve the bootstrapping problem.

**Comment by soatok** [comment]
For similar reasons that I made in https://github.com/fedi-e2ee/public-key-directory-specification/issues/118, I'd like to amend my previous recommendation for FEP-521a / RFC 9421 to include ML-DSA-44 keypairs by default instead of Ed25519.

(There isn't any real urgency for post-quantum security behind this amendment, I just want to minimize adoption pain when the industry is now accelerating towards post-quantum algorithms.)

**Comment by silverpill** [comment]
I think right now Ed25519 is a better option because it is better supported (and a significant number of Fediverse platforms already use it). 
Once tooling and standards catch up, we can migrate to ML-DSA-44.

**Comment by mariusor** [comment]
I think that at the moment RFC9421 does not actually have support for any of the quantum resistant algorithms: [current list](https://www.rfc-editor.org/rfc/rfc9421.html#iana-hsa-contents). 

So I think we first need someone with experience navigating the IANA submission mechanisms to push for their addition into the registry.

**Comment by soatok** [comment]
> So I think we first need someone with experience navigating the IANA submission mechanisms to push for their addition into the registry.

I'm on it already. I plan to tackle that this week. (Unfortunately, it's quite a busy month for me at work, so my goal might slip a bit timeline-wise.)

**Comment by gobengo** [comment]
> So I think we first need someone with experience navigating the IANA submission mechanisms to push for their addition into the registry.

I believe that's now done. See https://www.iana.org/assignments/http-message-signature/http-message-signature.xhtml

Also in general I think a change that would satisfy the OP of this issue, ie RFC9421 support, was resolved in https://github.com/mastodon/mastodon/pull/34814 and released in 4.5

**Comment by oneiros** [comment]
Actually, I think it was only resolved 3 days ago when we merged #39756.

As for ML-DSA-44 support, we have some already as of this week, but not yet for RFC9421 specifically as we are waiting for support in an upstream library.

In any case that is out-of-scope for this issue, so I am closing this now. Thank you everybody for chiming in.

**Comment by soatok** [comment]
> As for ML-DSA-44 support, we have some already as of this week, but not yet for RFC9421 specifically as we are waiting for support in an upstream library.

Please point me to the upstream library so I can help make that happen sooner than later.

**Comment by oneiros** [comment]
> > As for ML-DSA-44 support, we have some already as of this week, but not yet for RFC9421 specifically as we are waiting for support in an upstream library.
> 
> Please point me to the upstream library so I can help make that happen sooner than later.

https://github.com/nomadium/linzer/issues/28
